Feature/custom workspace backend

ericksantillanp opened the pull request:
<!--
Thank you for creating a pull request 🙌 ❤️
-->

**What kind of change does this PR introduce?**
<!-- (Update "[ ]" to "[x]" to check a box) -->

- [ ] Bugfix
- [x] New Feature
- [ ] Feature Improvement
- [ ] Refactoring
- [ ] Documentation
- [ ] Other, please describe:

**Description:**
<!--- Use this section to describe your changes.  Why is this change required? What problem does it solve? If your test fixes a specific issue, don't forget to reference the issue number. If your PR is still a work in progress, that's totally fine – just include a note to let us know. -->
This PR adds the possibility for a user to spawn a workspace with a different docker image than the one used by default. For this, I added an optional parameter `image` to the endpoint `/workspace/reset`. The behaviour is as follows:

* When the user doesn't specifies an image, it uses the default image.
* When the user specifies an image, this image is used to spawn the workspace. The volume under /workspace is mounted
* When the endpoint `/workspace/check` is called:
  * If the workspace has never been created, it creates one with the default image
  * If a workspace has already been created, it spawns a container with the latest image used

As a groundwork for an implementation in the front-end, I added information similar to the services in the API response (image name, container name, etc...)

**Checklist:**
<!--- Before you submit the PR, go over this checklist and make sure you can
tick off all the boxes. [] -> [x] -->

- [x] I have read the [CONTRIBUTING](https://github.com/sap/machine-learning-lab/blob/master/CONTRIBUTING.md) document.
- [x] My changes don't require a change to the documentation, or if they do, I've added all required information.

JNKielmann commented on backend/lab-service/src/main/java/org/mltooling/lab/api/LabAdminApiHandler.java:
```diff
@@ -237,9 +243,12 @@ public StatusMessageFormat checkWorkspace(String user) {
-      response.addMetadata(
-          "needsUpdate",
-          !CoreService.WORKSPACE.getImage().equalsIgnoreCase(workspaceService.getDockerImage()));
+      response.setData(workspaceService);
```
This will return a lot of information about the Service (for example all its environment variables).
The admin/workspace/check endpoint actually requires NO authentication.
So an malicious unauthenticated users can get the environment variables of the admin workspace which includes the admin JWT token and therefore get full admin privileges.
Suggestions:
- Create a separate class for the return value of this function instead of reusing the "LabService" class that contains too much sensitive data
- Add authorization to the admin/workspace/check endpoint so only admins can query it for any workspace and users can only query it for their own workspace

raethlein replied:
With regards to the suggestion of adding authorization to the endpoint, just keep in mind that it is also called from nginx and probably a change such as explicitly adding the user token to the request must be added.

JNKielmann replied:
Thanks for mentioning that. We already thought that there was probably a reason to not do authentication for the endpoint.
Alternatively, we could maybe leave the check workspace as it is (without authentication and only returning true/false) and add a new endpoint /workspace/info that returns workspace information and secure it properly.

raethlein replied:
yes, that could be an alternative. In case you want to try out the securing option, the line in nginx where the endpoint is called from Lua script is [here](https://github.com/SAP/machine-learning-lab/blob/9f1733776be8fc9ed894ce9b01bbfb382a72d0de/backend/lab-service/docker-res/nginx.conf#L247). Since you can access nginx variables there, I think you could access the lab-access token cookie / Auth Header there and set it in the lua request.

ericksantillanp replied:
Thanks for the heads-up and pointing us in the right direction @raethlein ! 
I added the token to the request [here](https://github.com/ericksantillanp/machine-learning-lab/blob/a8fcc3e74b2022efcc49da8868bbbabfd6b5ee1a/backend/lab-service/docker-res/nginx.conf#L249)

We also found out that when we use the authentication header with the token it shouldn't be passed as `Authorization: Bearer <TOKEN>` but rather `Authorization: <TOKEN>` . This is a little bit different than what is specified in the standard [RFC 6750](https://tools.ietf.org/html/rfc6750) for OAuth2. Is there a reason for this?

raethlein replied:
I don't think there is an intentional reason behind this and should probably better follow the RFC standard. Off the top of my head I don't know whether changing this will break anything or whether it must be changed at multiple places, so perhaps the safest solution would be to allow the "Bearer" keyword but not force it - or force it and then keep in mind that, if something breaks, this is the reason and must be added elsewhere.

JNKielmann commented on backend/environment-lib/src/main/java/org/mltooling/core/lab/LabAdminApiClient.java:
```diff
@@ -45,7 +45,7 @@ public String getEndpointUrl() {
+  public SingleValueFormat checkWorkspace(String workspaceId) {
     return executeRequest(
         Unirest.get(getEndpointUrl() + METHOD_CHECK_WORKSPACE)
             .getHttpRequest()
```
In line 53 (below) the type needs to be changed from StatusMessageFormat to SingleValueFormat

ericksantillanp replied:
Solved in [6838a86](https://github.com/SAP/machine-learning-lab/pull/26/commits/6838a86687c94f497d3e7cfaeb7f654c71c44c61)

JNKielmann commented on backend/lab-service/src/main/java/org/mltooling/lab/api/LabAdminApiHandler.java:
```diff
@@ -167,7 +168,13 @@ public StatusMessageFormat resetAllWorkspaces() {
    *     {WORKSPACE_SERVICE_PREFIX}
    */
-  public SingleValueFormat<LabService> resetWorkspace(String user) {
+  public SingleValueFormat<LabService> resetWorkspace(String user){
```
Code formatting seems wrong here. Make sure to follow the [google-java-format](https://github.com/google/google-java-format). There is an intelliJ plugin: https://plugins.jetbrains.com/plugin/8527-google-java-format/

ericksantillanp replied:
Solved in [6838a86](https://github.com/SAP/machine-learning-lab/pull/26/commits/6838a86687c94f497d3e7cfaeb7f654c71c44c61)

JNKielmann commented on backend/lab-service/src/main/java/org/mltooling/lab/endpoints/LabAdminEndpoint.java:
```diff
@@ -161,7 +163,7 @@ public Response resetAllWorkspaces(
       value =
           "Resets a workspace. Removes the container (keeps all persisted data) and starts a new"
-              + " one.",
+              + " one.  When specified an image, it is used this image instead",
```
Maybe reformulate for clarity: "If an image is specified, it will be used instead of the default image"

ericksantillanp replied:
Changed in [781b1d0](https://github.com/SAP/machine-learning-lab/pull/26/commits/781b1d017d0f46a8bc3a18289b453e10309ebaa2)

JNKielmann commented on backend/lab-service/src/main/java/org/mltooling/lab/services/AbstractServiceManager.java:
```diff
@@ -323,14 +328,24 @@ public boolean deleteWorkspace(String user) throws Exception {
+
+    if (previousUsedImage == null){
+      previousUsedImage = CoreService.WORKSPACE.getImage();
+      log.info("Unable to retrieve previous used image");
```
Do we need this log? If yes, we should also log the username for which this failed. Maybe make it a debug log?

ericksantillanp replied:
Changed in [781b1d0](https://github.com/SAP/machine-learning-lab/pull/26/commits/781b1d017d0f46a8bc3a18289b453e10309ebaa2)

JNKielmann commented on backend/lab-service/src/main/java/org/mltooling/lab/services/AbstractServiceManager.java:
```diff
@@ -323,14 +328,24 @@ public boolean deleteWorkspace(String user) throws Exception {
+
+    if (imageName == null){
+      imageName = previousUsedImage;
+      log.info("Not image given, used previous used image: " + imageName);
```
Do we need this log? If so I would make it a debug log and maybe reformulate to:
"No image provided, using previously used image."

ericksantillanp replied:
Changed in [781b1d0](https://github.com/SAP/machine-learning-lab/pull/26/commits/781b1d017d0f46a8bc3a18289b453e10309ebaa2)

ericksantillanp commented:
I added the token to the nginx request as pointed out before.
I added some integration tests for the custom workspace. You can see them [here](https://github.com/ericksantillanp/machine-learning-lab/blob/feature/custom-workspace-backend/backend/lab-service/src/test/java/org/mltooling/lab/LabAdminApiTest.java#L216). This also showed me that there were some modifications to be made on the Java client:  
* Change signature of the checkWorkspace
* Add parameter to resetWorkspace

JNKielmann reviewed (approved):
I looked through the new changes and successfully ran the tests you have added. It looks good to me👍
